src/magic_folder/test/test_magic_folder_service.py

```diff
@@ -117,6 +148,182 @@ def test_create_local_snapshot(self, relative_target_path, content):
         )
 
         self.assertThat(
-            snapshot_creator.processed,
+            local_snapshot_creator.processed,
             Equals([target_path]),
         )
+
+    def test_start_uploader_service(self):
+        """
+        When the ``MagicFolder`` service is started the given uploader service is
+        also started.
+        """
+        magic_path = FilePath(self.mktemp())
+        magic_path.asBytesMode("utf-8").makedirs()
+
+        local_snapshot_creator = LocalMemorySnapshotCreator()
+        local_snapshot_service = LocalSnapshotService(magic_path, local_snapshot_creator)
+        clock = task.Clock()
+
+        # create RemoteSnapshotCreator and UploaderService
+        uploader_service = Service()
+
+        tahoe_client = object()
+        name = u"local-snapshot-service-test"
+        config = object()
+        participants = object()
+        magic_folder = MagicFolder(
+            client=tahoe_client,
+            config=config,
+            name=name,
+            local_snapshot_service=local_snapshot_service,
+            uploader_service=uploader_service,
+            initial_participants=participants,
+            clock=clock,
+        )
+        magic_folder.startService()
+        self.addCleanup(magic_folder.stopService)
+
+        self.assertThat(
+            uploader_service.running,
+            Equals(True),
+        )
+
+
+LOCAL_AUTHOR = find(local_authors(), lambda x: True)
+
+class MagicFolderFromConfigTests(SyncTestCase):
+    """
+    Tests for ``MagicFolder.from_config``.
+    """
+    @given(
+        folder_names(),
+        relative_paths(),
+        path_segments(),
+        relative_paths(),
+        just(LOCAL_AUTHOR),
+        one_of(
+            tahoe_lafs_dir_capabilities(),
+            tahoe_lafs_readonly_dir_capabilities(),
+        ),
+        tahoe_lafs_dir_capabilities(),
+        integers(min_value=1, max_value=10000),
+        binary(),
+    )
+    def test_uploader_service(
+            self,
+            name,
+            file_path,
+            relative_magic_path,
+            relative_state_path,
+            author,
+            collective_dircap,
+            upload_dircap,
+            poll_interval,
+            content,
+    ):
+        """
+        ``MagicFolder.from_config`` creates an ``UploaderService`` which will
+        upload snapshots using the given Tahoe client object.
+        """
+        reactor = task.Clock()
+
+        root = create_fake_tahoe_root()
+        http_client = create_tahoe_treq_client(root)
+        tahoe_client = TahoeClient(
+            DecodedURL.from_text(U"http://example.invalid./"),
+            http_client,
+        )
+
+        root._uri.data[upload_dircap] = json.dumps([
```

Comment:
`_FakeTahoeRoot` has an `add_data` method which is probably better than twiddling internals like this directly.

I guess here you're doing something a bit different, though, which is adding data at an explicit key. The `add_data` API returns you the key (like Tahoe would). So a way around that would be to get rid of the Hypothesis generator for the upload-dircap and just generate it in the test like `_, upload_dircap = root.add_data("URI:DIR2-CHK", json.dumps(...))` (approximately).